lib/sarif/base_sarif.rb

```diff
@@ -0,0 +1,146 @@
+          "rules" => rules
+        }
+      }
+    end
+
+    # Retrieves result section for sarif report
+    def build_result(parsed_issue)
+      result = {
+        "ruleId": parsed_issue[:id],
+        "ruleIndex": @mapped_rules[parsed_issue[:id]],
+        "level": sarif_level(parsed_issue[:level]),
+        "message": {
+          "text": parsed_issue[:details]
+        },
+        "locations": [
+          {
+            "physicalLocation": {
+              "artifactLocation": {
+                "uri": parsed_issue[:uri],
+                "uriBaseId": "%SRCROOT%"
+              }
+            }
+          }
+        ]
+      }
+      if parsed_issue[:code]
+        result[:locations][0][:physicalLocation][:region] = {
+          "startLine": parsed_issue[:start_line].to_i,
+          # "endLine": parsed_issue[:end_line.to_i]
+          "snippet": {
+            "text": parsed_issue[:code]
+          }
+        }
+      end
+      result
+    end
+
+    def build_rule(parsed_issue)
+      if !@mapped_rules.include?(parsed_issue[:id])
+        rule = {
+          "id": parsed_issue[:id],
+          "name": parsed_issue[:name],
+          "fullDescription": {
+            "text": parsed_issue[:details]
+          },
+          "helpUri": parsed_issue[:help_url] || '',
+          "help": {
+            "text": "More info: #{parsed_issue[:help_url]}",
+            "markdown": "[More info](#{parsed_issue[:help_url]})."
+          }
+        }
+        @mapped_rules[parsed_issue[:id]] = @rule_index
+        @rule_index += 1
+        if rule[:id] == 'SAL0002'
+          rule[:fullDescription][:text] = 'Golang errors generated at runtime'
+        end
+        rule
+      end
+    end
+
+    # Retrieves invocation object for SARIF report
+    def build_invocations
+      {
+        "executionSuccessful": @scan_report.passed? || false,
+        "toolExecutionNotifications": [{
+          "descriptor": {
+            "id": "SAL001"
+          },
+          "message": {
+            "text": "SARIF reports are not available for this scanner"
+          }
+        }]
+      }
+    end
+
+    # Returns the 'runs' object for the scanners report
+    def build_runs_object
+      results = []
+      rules = []
+      @logs.each do |issue|
+        parsed_issue = parse_issue(issue)
+        rule = build_rule(parsed_issue)
+        rules << rule if rule
+        results << build_result(parsed_issue)
+      end
+
+      {
+        "tool" => build_tool(rules: rules),
+        "conversion" => build_conversion,
+        "results" => results,
+        "invocations" => [build_invocations]
+      }
+    end
+
+    # Returns the conversion object for the SARIF report
+    def build_conversion
+      {
+        "tool": {
+          "driver": {
+            "name": "Salus",
+            "informationUri": DEFAULT_URI
+          }
+        }
+      }
+    end
+
+    # Returns a sarif wraning level for a given severity
+    def sarif_level(severity)
```

Comment:
How did you decide on the severity to level mappings?
Also, we need separate keys for severity and confidence.

Reply:
In the [SARIF document](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.pdf), there are defined fields for different objects. For the result object(@3.27), there was no mention about confidence or severity. The closest option for these two is the level object(@3.27.10). If i add severity and confidence fields the program would raise a SarifInvalidFormatError because it would not follow the schema. The mappings were done based on how gosec currently generates its levels from its severity. https://github.com/securego/gosec/blob/master/output/sarif_format.go#L173